### PR TITLE
Non-relative includes, restructure, add missing includes

### PR DESCRIPTION
**Pull request by NeroBurner**
This is a two-part PR, first use relative includes, and secondly restructure all includes.

The non-relative includes are needed for me in #743 to inject my header-stubs for the simulator and is a more robust coding style for header files

---

First commit message:

Don't use relative imports like `../foo.h` as those depend on the
relative position of both files. Rather than that use imports relative
to the `src` directory, which explicitly is part of the include
directories.

---

Second commit message:

Update the include structure:
- use `#include "foo.h"` for headers from the InfiniTime project (except
  bundled libraries)
- use `#include <foo.h>` for everything else (external libraries or standard
  library)

Change the include order from the most project specific down to the most
generic include. This helps to enforce, that our headers include what
they use. For example if `foo.h` uses `std::array` but forgets the
include. This won't be cought if `main.cpp` includes `<array>` before
`foo.h`.

For example `DeviceInformationService.h` did miss `<stdint>` include for
`uint16_t` and `BootloaderVersion.h` missed `<stddef>` for `size_t`.
As last example `InfiniPaint.cpp` missed include `<algorithm>` for `std::fill`.

The following include order is used:
- project headers
- nimble library
- littlefs library
- date library
- lvgl library
- FreeRTOS library
- nrf5 sdk
- standard libraries

### CODE REVIEW

File: src/components/alarm/AlarmController.cpp
Comment (Avamander) [inline src/components/alarm/AlarmController.cpp:24]: I don't think the comments are necessary.

Comment (NeroBurner) [reply src/components/alarm/AlarmController.cpp:24]: They add a bit of structure for future contributions to keep the include order. And it would have been helpful for my first exploration of the source code. For example I had no idea, that `<task.h>` is part of FreeRTOS. With the comment this is explicitly listet

**Comment by Avamander** [comment]
Includes were sorted once a while ago, I don't see a reason to redo that part with the potential downside that InfiniTime's code conflicts with any of the libraries'.

Original sort order
```
Sort includes:
* source files include their header files first
* library includes next
* own code includes last
* alphabetical
```

File: src/components/alarm/AlarmController.cpp
Comment (Avamander) [reply src/components/alarm/AlarmController.cpp:24]: Sounds like a dev documentation addition instead of each file.

**Comment by NeroBurner** [comment]
One advantage of the new ordering is, that I was able to find some missing includes, that were hidden by the previous ordering.

I don't understand the comment about conflicting library code, all the targets still compile. Would you please be so kind to explain this point to me so I can understand as well?

**Comment by Avamander** [comment]
> I don't understand the comment about conflicting library code, all the targets still compile. Would you please be so kind to explain this point to me so I can understand as well?

Yes, at the moment. There'd be a potential for conflicts or nasty bugs when we define something and it ends up being used in a library, because ours comes first instead of being last.

**Comment by NeroBurner** [comment]
I think we won't (and shouldn't) define so short and generic macros, that they get picked up in libraries. Those changes would be bad code for me (personally).

Even if that happens we could move those specific library includes up and still get the "include-what-you-use"-checkings

Another way to argue is, that library defines could bleed into our headers and subtly break stuff. I think the argument works in both directions

---

sorry for being pedantic about this. I just think it is "more right" and more resilient with the new ordering. In the end it's your project and your guidelines and I'll adhere to them (after a proper discussion :grin: )

**Comment by Avamander** [comment]
> Another way to argue is, that library defines could bleed into our headers and subtly break stuff. I think the argument works in both directions

Yes, but from the direction of libraries to us it's less error prone.

**Comment by Avamander** [comment]
> I think we won't (and shouldn't) define so short and generic macros, that they get picked up in libraries. Those changes would be bad code for me (personally).

Yes absolutely.

Basically the thing is that I see a very minor upside to not changing the order (less changes + our code affects only our own), but I think e.g. @geekbozu or @Riksu9000 should break the tie here.

**Comment by Riksu9000** [comment]
I would agree with @Avamander here. I don't see enough reason to change it.

File: src/components/alarm/AlarmController.cpp
Comment (Riksu9000) [inline src/components/alarm/AlarmController.cpp:18]: Do these cpp file specific headers also need to be non relative? Normally it's pretty safe to assume they're in the same directory.

Comment (NeroBurner) [reply src/components/alarm/AlarmController.cpp:18]: not strictly necessary, just consistent

**Comment by NeroBurner** [comment]
reverted ordering commit, readd the found missing standard includes

File: src/components/alarm/AlarmController.cpp
Comment (Riksu9000) [reply src/components/alarm/AlarmController.cpp:18]: I would personally rather not type the entire path for these cases as it seems unnecessary. For everything else it's fine.

**Comment by Riksu9000** [comment]
Since we're already restructuring, maybe the different sections should be separated by an empty line, which would make the sorting order more obvious and then clang-tidy wouldn't suggest sorting them in a different way.

File: src/components/alarm/AlarmController.cpp
Comment (NeroBurner) [reply src/components/alarm/AlarmController.cpp:18]: @geekbozu @Riksu9000 what's your stance on this?

- :+1: for consistently using `#include "components/alarm/AlarmController.h"`
- :-1: for using the shorter `#include "AlarmController.h"` (IMHO looks inconsistent as soon as a include from the same component is directly after the main header include)

Comment (Riksu9000) [reply src/components/alarm/AlarmController.cpp:18]: I have been trying to think about which would be better, but honestly I'm fine with either. If no one has anything to add, then we can just leave it as it is.

**Comment by Riksu9000** [comment]
If every include must be non relative, it should be mentioned here https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/doc/contribute.md#coding-style

File: src/components/alarm/AlarmController.cpp
Comment (Avamander) [reply src/components/alarm/AlarmController.cpp:18]: I don't think we'll end up with name conflicts, unless it's something generic. But I guess it wouldn't hurt to use the full path

Comment (geekbozu) [reply src/components/alarm/AlarmController.cpp:18]: I'm a fan of verbosity, while it is a pain to have them long...its only a few more keystrokes for explicitness, And explicit saves headache later.

**Comment by NeroBurner** [comment]
moving up the file header include to the top, one example from `WatchFaceAnalog.cpp`:

```cpp
#include "displayapp/screens/WatchFaceAnalog.h"
#include <lvgl/lvgl.h>
#include "displayapp/screens/BatteryIcon.h"
#include "displayapp/screens/BleIcon.h"
#include "displayapp/screens/Symbols.h"
#include "displayapp/screens/NotificationIcon.h"
```

Is this OK? Or should there be a space before and after the `lvgl/lvgl.h` include?

**Comment by NeroBurner** [comment]
> If every include must be non relative, it should be mentioned here https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/doc/contribute.md#coding-style

@Riksu9000 is this Ok? https://github.com/InfiniTimeOrg/InfiniTime/pull/805/commits/0d745a0ad10378d71196af908f031108252b1214

**Review by Avamander — APPROVED** [review]
Seems fine now. Someone should take a look as well.

**Comment by Avamander** [comment]
Oh no, a conflict, if you fix that and we can merge, we can avoid breaking this PR with each and every other PR.

**Comment by NeroBurner** [comment]
rebased and conflict resolved

**Review by Riksu9000 — APPROVED** [review]

**Comment by NeroBurner** [comment]
rebased on current develop (after v1.7.0 release)

**Comment by NeroBurner** [comment]
rebased on `develop` after v1.7.1 release

**Comment by NeroBurner** [comment]
@Avamander Is something still missing, did I forget something? Or is it just limited free time for this awesome hobby? :)

**Comment by Avamander** [comment]
Everything seems to be in order, I think it's in the queue waiting for a merge.

**Comment by JF002** [comment]
@NeroBurner 

> Is something still missing, did I forget something? Or is it just limited free time for this awesome hobby? :)

I'm sorry for the time it took for me to have a look at your PR! I am currently the only one that is allowed to merge PRs to develop (that might change in the future) and I couldn't allocate much time to InfiniTime recently!

However, I think you did a great job cleaning the includes of the project! Thank you for that! 
I've just fixed the last conflicts that were introduce by PRs I've just merged and I merged this one!

**Comment by NeroBurner** [comment]
@JF002 Thanks :tada: Happy it's merged

InfiniTime is a hobby-project. If there is no personal time left to work on it, then it is ok as well ;) It's true, that waiting for feedback for weeks was bad for motivation, but now I'm happy my contribution got merged.

I'm VERY grateful to you, that you created InfiniTime. Better take some personal time, than burning out on the project

In short: thank you @JF002  :)

**Comment by JF002** [comment]
@NeroBurner Thank you so much for your kind words! Do not forget I'm not alone in this adventure! There are so many contributors and enthousiasts and users that supports the project! Thank to all of them!
I understand that waiting several weeks for a feedback is not ideal, and that's exactly for that reason that I'm creating teams of contributors that help me to triage the issues, PRs, review the code, help with specific topics,...
Thanks again, and have fun contributing again to the project :-)

**Pull request merged**
